src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackWriter.cs

```diff
@@ -1210,7 +1214,7 @@ private static void WriteBigEndian(ulong value, Span<byte> span)
             // solves heuristic length check
 
             // ensure buffer by MaxByteCount(faster than GetByteCount)
-            bufferSize = StringEncoding.UTF8.GetMaxByteCount(characterLength) + 5;
+            bufferSize = (characterLength * 3) + 5;
```

Comment:
Where did you get this `3` multiplier? UTF8's longest encoded character is 4 [according to this](https://stackoverflow.com/questions/9533258/what-is-the-maximum-number-of-bytes-for-a-utf-8-encoded-character).  
As it is, the `GetMaxByteCount` method looks relatively simple so I don't see how this is a significant perf win. A general rule for accuracy vs. performance is "[as simple as possible *but no simpler*.](https://championingscience.com/2019/03/15/everything-should-be-made-as-simple-as-possible-but-no-simpler/#:~:text=%E2%80%9CEverything%20should%20be%20made%20as%20simple%20as%20possible%2C,a%20great%20statement%20of%20how%20to%20conduct%20science.)" Based on the [.NET implementation of this method](https://source.dot.net/#System.Private.CoreLib/UTF8Encoding.cs,199682d3d0340ec9) it looks like you're over-simplifying this, potentially at the risk of under-estimating the maximum required buffer size.

Reply:
Within the BMP, UTF8 requires a maximum of three bytes; UTF16 requires one character.
Outside the BMP range, UTF16 uses two characters; UTF8 always requires four bytes.
That's two bytes UTF8 per UTF16 character.

So it's estimated to be up to 3 bytes UTF8 per one character UTF16.

Reply:
.NET implementation is too complex because they want to handle the invalid UTF16 strings.
It is undefined how to encode the invalid UTF16 to valid UTF8 bytes.

Reply:
What is BMP? And can you point to an RFC (or another doc of similar notoriety) to back up what you're saying? I really don't want to regress this and the first RFC I found said 4 was the magic number.

Reply:
> Also: did you actually see GetMaxByteCount taking too long in a perf trace? 
No, I didn't so I agree with you.